sycl/source/detail/online_compiler/online_compiler.cpp

```diff
@@ -0,0 +1,208 @@
+    int CurrentVersionMajor = (ocloc_version_t::OCLOC_VERSION_CURRENT) >> 16;
+    if (LoadedVersionMajor != CurrentVersionMajor)
+      throw online_compile_error(
+          std::string("Found incompatible version of ocloc library: (") +
+          std::to_string(LoadedVersionMajor) + ", " +
+          std::to_string(LoadedVersion & 0xffff) +
+          "). The supported versions are (" +
+          std::to_string(CurrentVersionMajor) + ", *).");
+
+    CompileToSPIRVHandle =
+        sycl::detail::pi::getOsLibraryFuncAddress(OclocLibrary, "oclocInvoke");
+    if (!CompileToSPIRVHandle)
+      throw online_compile_error("Cannot load oclocInvoke() function");
+    FreeSPIRVOutputsHandle = sycl::detail::pi::getOsLibraryFuncAddress(
+        OclocLibrary, "oclocFreeOutput");
+    if (!FreeSPIRVOutputsHandle)
+      throw online_compile_error("Cannot load oclocFreeOutput() function");
+  }
+
+  std::string CombinedUserArgs;
+  for (auto UserArg : UserArgs) {
+    if (UserArg == "")
+      continue;
+    if (CombinedUserArgs != "")
+      CombinedUserArgs = CombinedUserArgs + " " + UserArg;
+    else
+      CombinedUserArgs = UserArg;
+  }
+  std::vector<const char *> Args = detail::prepareOclocArgs(
+      DeviceType, DeviceArch, Is64Bit, DeviceStepping, CombinedUserArgs);
+
+  uint32_t NumOutputs = 0;
+  byte **Outputs = nullptr;
+  size_t *OutputLengths = nullptr;
+  char **OutputNames = nullptr;
+
+  const byte *Sources[] = {reinterpret_cast<const byte *>(Source.c_str())};
+  const char *SourceName = "main.cl";
+  const uint64_t SourceLengths[] = {Source.length() + 1};
+
+  Args.push_back("-file");
+  Args.push_back(SourceName);
+
+  decltype(::oclocInvoke) *OclocInvokeFunc =
+      reinterpret_cast<decltype(::oclocInvoke) *>(CompileToSPIRVHandle);
+  int Error =
+      OclocInvokeFunc(Args.size(), Args.data(), 1, Sources, SourceLengths,
+                      &SourceName, 0, nullptr, nullptr, nullptr, &NumOutputs,
+                      &Outputs, &OutputLengths, &OutputNames);
+
+  byte *SpirV = nullptr;
+  std::string CompileLog;
+  size_t SpirVSize = 0;
+  for (uint32_t I = 0; I < NumOutputs; I++) {
+    size_t NameLen = strlen(OutputNames[I]);
+    if (NameLen >= 4 && strstr(OutputNames[I], ".spv") != nullptr) {
+      SpirVSize = OutputLengths[I];
+      SpirV = new byte[SpirVSize];
+      std::memcpy(SpirV, Outputs[I], SpirVSize);
+    } else if (!strcmp(OutputNames[I], "stdout.log")) {
+      CompileLog = std::string(reinterpret_cast<const char *>(Outputs[I]));
+    }
+  }
+
+  if (Error)
+    throw online_compile_error("ocloc reported compilation errors {" +
+                               CompileLog + "}");
+
+  decltype(::oclocFreeOutput) *OclocFreeOutputFunc =
+      reinterpret_cast<decltype(::oclocFreeOutput) *>(FreeSPIRVOutputsHandle);
+  Error =
+      OclocFreeOutputFunc(&NumOutputs, &Outputs, &OutputLengths, &OutputNames);
+
+  if (Error)
+    throw online_compile_error("ocloc cannot safely free resources");
```

Comment:
SpirV not freed on error

Reply:
In this case, I don't feel safe freeing resources right here and duplicating/doing oclocFreeOutput() job

Reply:
ok, makes sense